### DISCUSSION TITLE
【AudioMNIST-diffusion】So many modules

### DISCUSSION THREAD

**Discussion by a897456** [speechbrain/speechbrain#2612] [General]
https://github.com/speechbrain/speechbrain/blob/47dd9ece5363f4410fa560da1cc2b850a8cb44fd/recipes/AudioMNIST/diffusion/hparams/train.yaml#L214-L225
There are so many modules. I don't know which module to use exactly? Or are all of them to be used? I intended to understand through the code, but failed. Can someone help to explain?

**Comment by a897456** [comment]
https://github.com/speechbrain/speechbrain/blob/47dd9ece5363f4410fa560da1cc2b850a8cb44fd/recipes/AudioMNIST/diffusion/train.py#L113-L121
Should  `self.optimizer` in line 119 be replaced by `self.optimizer_done`？

**Reply by asumagic** [reply]
Not sure. What do you think @flexthink ?

**Reply by flexthink** [reply]
That is correct - the original recipe has a typo. While it won't prevent the model from training, it needs to be fixed

**Comment by asumagic** [comment]
All of those modules are used. The `modules` key in the yaml groups up most modules (usually to do grouped actions like moving the modules to CUDA).  
It's not particularly useful to understand how the model works.

To understand how the model works you generally want to look at the contents of `compute_forward` and `compute_objective` methods step by step, while also looking at hyperparameters to see what they refer to.

Looking up tutorials/articles/papers on the relevant architecture may also be helpful, as the source code is sometimes not written or documented in a way that is helpful to understand the basics of the architecture.
